pkg/restore/restore.go

```diff
@@ -1358,14 +1366,6 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
 	// and which backup they came from.
 	addRestoreLabels(obj, ctx.restore.Name, ctx.restore.Spec.BackupName)
 
-	if ctx.resourceModifiers != nil {
-		if errList := ctx.resourceModifiers.ApplyResourceModifierRules(obj, groupResource.String(), ctx.log); errList != nil {
-			for _, err := range errList {
-				errs.Add(namespace, err)
-			}
-		}
```

Comment:
I personally feel it's better if we do changes with the new namespace, (resource modifier filter having new namespace). Since this will lead to less confusion IMO. And also easier for the end user to debug any issues in resourcemodifier setup.

Otherwise think of the scenario where the original namespace is already there in the cluster and wishes to restore to a diff namespace, IMO the filters should target the new namespace for clarity.

Reply:
We can also document this behaviour in docs.

Reply:
But I think the goal of resource modifier is to modify an object from old to new, so there is a mapping in resource modifier config(an old part and a new part). To me, name, namespace, label and old value in given path belong to the old part, and new value is the new part. If I want to restore an object with resource modifier, I will first write rules according the origin object and pass it to velero to perform the restore. I might later change my mind and want to change the namespace or not, the rules should remain unchanged.

Reply:
so namespace is just part of the change which happens if you see.

There are the other RestoreItemAction plugins which can change the YAMLs based on various rules.  Even there could be RIA to change the namespace based on certain rules.

In the design doc the approach we took was to modify just before creation and after all RIA / modifications are applied.

Our docs line could indicate that both namespace mapping and RIA would have run before Resourcemodifier is applied.

I think of RIA and namespace mapping together, either we do resourcemodifier before both or after both

In between ismore confusing IMO.

Reply:
I prefer to do resourcemodifier before both

Reply:
The issue is that there are certain RIA say CSI plugin ones which are dependent on multiple resources -> VSC + PVC 

My worry is that resourcemodifier changes can potentially break the behavior of such plugins / RIAs, and it will be very hard to debug for end customer such scenarios.

Reply:
why do you think "In between ismore confusing"?

Reply:
because RIA can also potentially change namespace and namespace mapping can also change namespace

Reply:
So this PR only affect users who add resourceModifier for namespace field correct? If so, until namespace mapping feature is replaced by resourceModifier, if it ever is, resourceModifier shouldn't apply on namespace values IMO. It's meant for other spec fields.

Reply:
Namespace mapping is handled separately -- it would be very confusing if the resource modifier code used both old and new ns (one on each side). Instead, we should pick which one gets used (backup namespace, or mapped-in-restore namespace), document it clearly, and stick with it. I don't have a strong opinion on which one we use, as long as it's clear, but we don't want to use *both* here.

Reply:
@kaovilai "If so, until namespace mapping feature is replaced by resourceModifier" -- I don't see any reason to ever consider this. NS mapping is a separate feature that determines *where* resources go, this feature (as well as RIA/BIA actions) determine *what those resources look like*.

Reply:
@anshulahuja98 "because RIA can also potentially change namespace" -- I suspect that if a RIA did this, it would result in errors. We use the target namespace from the NS mappings to pass into the dynamic client used for object creation. While it's not documented clearly right now, we probably want to document that neither resource modifier nor RIA namespace changing is supported.

Reply:
@sseago I agree with you that a RIA should never change namespace, here is a bad example: https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/82

Reply:
Some reasons for using old namespace:
1. resource modifier is to modify an object from old to new, old object is more clear since new object can be changed.
2. resource modifier is a config (ConfigMap), and may be a large one, it should be reusable when some restore options are changed, like namespace mapping

Reply:
@27149chen I think your second reason is the biggest reason to stick with the old namespace. Logically, namespace mapping is one of the last things to do -- first we change *what* we restore, then we change *where* we restore it to. And yes, you might have one backup that you're using as a template for restoring to different namespaces. Basically for any given backup, the old namespace will be valid for every restore that's made from it, but the new namespace may not.

Reply:
@27149chen So the CSI example seems like a bit of a special case. It's not arbitrarily changing namespace outside of the NS mapping -- if it were to do that, I suspect that it would break things. Instead, it's actually using the NS mapping and changing it here (same change as velero would eventually change without it), but the reason for it is that the namespace is then used to find the matching VolumeSnapshot resource, so it needs to use the new namespace here.

Reply:
@27149chen  can you update the website docs and resource modifier design docs with this callout for changes done before NS mapping?
Post that we can get this checked in. Since overall folks are inline with your proposal of using old namespace, I agree with further comments and the majority.